### DISCUSSION TITLE
YTuner testing cases.

### DISCUSSION THREAD

**Discussion by coffeegreg** [coffeegreg/YTuner#32] [Tests]
Hello everyone. :hand: 

This discussion is created for testing new versions of YTuner before their final release. :+1: 
As you know, the specificity of this project means that I cannot test everything at home because I do not own all the AVRs that YTuner should work with. That's why I'm asking you for help with testing. :wink: 

First, an idea for bypassing the problem related to issue #28 . It assumes the ability to add an empty "dummy" folder (in this case the folder about the YTuner service) and additionally allows users to give their own names to existing folders such as `My Stations`, `Bookmark` and `Radio Browser` and change their order in the main menu.
For example, now you can change the name of "My Stations" to the correct one for your language :
- "Moje stacje radiowe" (pol)
- "Meine Radiosender" (ger)
- "Mes radios" (fr)

... or any other, paying attention to whether your AVR supports your national characters.

Everything is set via two new parameters in the `avr.ini` file (if you use one common file for all your devices) or in an analogous ini file dedicated to the selected AVR device. Here they are with their description:

>[MainMenu Items]
; This setting is critical for properly creating your AVR's main menu! Items have to be separetad by ";"
;
; IdentifiersList : supported values :
; - mystations : AKA "My Stations". Contains stations from stations.ini or stations.yml file(s).
; - radiobrowser : AKA "Radio browser". Contains stations from Radio Browser (https://www.radio-browser.info/).
; - bookmarks : AKA "Bookmark". Contains stations added from your AVR's remote control.Your AVR must support bookmarks to save stations to bookmark.
; - about : "Dummy" item meaning empty folder with "welcome" info from YTuner. Useful sometimes when your AVR device handles a specific menu item in a special way, e.g. the first one. (Look at this issue : https://github.com/coffeegreg/YTuner/issues/28)
; DON'T USE any other values! They won't work!
; All those values can be disabled by corresponding settings from ytuner.ini file! (Read ytuner.ini file description for: [MyStations]:Enable, [Radiobrowser]:Enable, [Bookmark]:Enable)
;
; LabelsList : list of corresponding labels displayed on AVR.
;
; The number of elements separated by ";" in the LabelsList MUST be the same as the number of elements in IdentifiersList.
; Hint: You can not use ";" character in element name because this character cast as separator!
;
; Setting examples :
;  - AVR support bookmarks :
;    IdentifiersList=mystations;radiobrowser;bookmarks
;    LabelsList=My Favourites;Radio Browser;Bookmarks
;
;  - AVR doesn't support bookmarks :
;    IdentifiersList=mystations;radiobrowser
;    LabelsList=My Favourites;Radio Browser
;
;  - AVR with a "special" handle of first menu item and bookmarks support :
;    IdentifiersList=about;mystations;radiobrowser;bookmarks
;    LabelsList=# YTuner #;My Favourites;Radio Browser;Bookmarks
;
;  - Some Yamaha AVRs with bookmarks support (bookmarks originally as the first menu item) :
;    IdentifiersList=bookmarks;mystations;radiobrowser
;    LabelsList=Bookmarks;My Favourites;Radio Browser
;
IdentifiersList=about;bookmarks;mystations;radiobrowser
LabelsList=# YTuner #;Favourites;My Stations;Radio Browser

YTuner will automatically enter default values ​​for these parameters and update the avr.ini file version.
Change them as you wish but stick to the rules described.

Additionally, the version includes a fix in which YTuner sends the entire contents of the folder (instead of just the first one) if the AVR device does not provide the range it is interested in (`start` , `howmany` or `startitems` , `enditems`) .

Finally, some most popular binaries  of `YTuner v1.2.4 alpha` to test :rocket: 

[ytuner-1.2.4-alpha-aarch64-linux.zip](https://github.com/user-attachments/files/17181301/ytuner-1.2.4-alpha-aarch64-linux.zip)
[ytuner-1.2.4-alpha-arm-linux.zip](https://github.com/user-attachments/files/17181302/ytuner-1.2.4-alpha-arm-linux.zip)
[ytuner-1.2.4-alpha-x86_64-linux.zip](https://github.com/user-attachments/files/17181303/ytuner-1.2.4-alpha-x86_64-linux.zip)

**Comment by TheBossME** [comment]
Hi Greg,

good idea 🙏 to change the Menu items.
How about inside the radio-browser.info (Country=Land) etc ?
is it possible to change automatically while using language specific setting ?. This technique is used by vtuner.com

![image](https://github.com/user-attachments/assets/0abb2684-3227-424e-894d-ef6aced7193b)

![image](https://github.com/user-attachments/assets/6e2c06ce-c9fb-4ec6-b2ab-ae3cd13d4775)

![image](https://github.com/user-attachments/assets/3353062a-db36-40b8-9c24-c91cf5213260)

**Comment by coffeegreg** [comment]
Hi Beatrice,

We can always count on your help. :clap:
>How about inside the radio-browser.info (Country=Land) etc ?
is it possible to change automatically while using language specific setting ?. This technique is used by vtuner.com

Yes, I can do that but please note that vtuner has and maintains its own station database and manages category names itself and can therefore support multiple languages. YTuner uses `radio-browser.info` and we have no influence on how the data is implemented there now and in the future. I can enable customization of display labels: "Genre", "Countries", "Languages" and "Most popular", but what is deeper in these folders must remain as it was entered in the radio-browser.info database. Theoretically, we could make dictionaries that would try to translate on the fly, but I don't think it would make much sense. :thinking: 
But... let's get back to the tests.
What do you see on your Yamaha display/TV screen when you try to enter one of those "dummy" folders like `Welcome to YTuner v1.2.4 alpha` or `Copyright ...` You should see something like `[No content]` or `This folder is intentionally empty...`.
The question concerns the ability of our AVRs to display messages sent by YTuner in the form of
>
	<Item>
		<ItemType>Display</ItemType>
		<Display>This folder is intentionally empty...</Display>
	</Item>

YTuner supports such messages but neither my RX-V671 nor RX-V673 support them. :disappointed: 
I wonder if your DSP-Z7 will display them. :thinking: 

I'm also waiting for @arun-doss to test this alpha version. He promised to help. :+1: 
Anyway... is anyone interested? :eyes:

**Reply by TheBossME** [reply]
This is what I see.

![image](https://github.com/user-attachments/assets/6276b6c1-8efb-45f8-967e-dff284df62a0)

![image](https://github.com/user-attachments/assets/cfd1e2c0-c03c-45b5-a671-d6cd52019f07)

**Comment by coffeegreg** [comment]
Yes, DSP-Z7 also does not support display messages :confused:
No problem.. it's nothing important really. But how pure the sound of our Yamaha AVRs is. :notes: :+1: :smiley:

**Reply by TheBossME** [reply]
Attachment 1.2.3.5+ ?
I am blind ?

**Reply by coffeegreg** [reply]
This mean 1.2.3+++++ :wink: For testing purposes only. Look here #28.

**Comment by arunprasaddoss** [comment]
For my AVR RX-V677 everything works seamless now. 
@coffeegreg  thanks a lot for taking your time for supporting my AVR.

**Comment by coffeegreg** [comment]
Hello to all my :star: [Stargazers](https://github.com/coffeegreg/YTuner/stargazers) :star:

`YTuner v1.2.4 beta 1` is ready to test. :rocket: 
- Fix : YTuner resolves the original vTuner radio station links and redirects to the URL of the correct radio station. (Fix to this issue : https://github.com/coffeegreg/YTuner/issues/33 - Thank you @Stijn-Daniels )
- Fix : YTuner returns a full list of stations even when the AVR does not transmit a range (Fix to this issue: https://github.com/coffeegreg/YTuner/issues/28  - Thank you @arun-doss)
- Improvement : Some improvement described here (https://github.com/coffeegreg/YTuner/discussions/32#discussion-7253563)
- Improvement : Translator for non-English users based on Google Translate service. Beware of Google Translate Service limitations :exclamation:  (Inspired by Beatrice @TheBossME : https://github.com/coffeegreg/YTuner/discussions/32#discussion-7253563)
- Improvement : Possibility to replace diacritical characters defined in "UTF-8 Latin1 Supplement" and/or "UTF-8 Latin Extended A" and/or "UTF-8 Latin Extended B". You can read about this [here](https://www.w3schools.com/charsets/ref_html_utf8.asp). Use it if your AVR cannot display your national characters.

Example of new `avr.ini` settings :
```
[Translator]
; Enables translator for selected language for all menus/submenus items releated to RadioBrowser submenu. Radio station names will remain untranslated! 
; Translator file is a simple [name]=[value] pairs file where [name] is original word or phrase (English in most cases) and [value] is tranlation for a selected laguage;
; Your local translation files will be created and systematically populated based on what you browse, but without the translations themselves until you add them yourself or enable "Auto" setting (see below).
; Example (Polish):
; Classic=Klasyczna
; Languages=Języki
; Countries=Kraje
; Example (German):
; Classic=Klassiker
; Languages=Sprachen
; Countries=Länder
; Warning! Make sure your AVR supports your national characters! Otherwise, use the settings below: ReplaceUTF8Latin1Supplement, ReplaceUTF8Latin1ExtendedA, ReplaceUTF8Latin1ExtendedB.
; 1-Enabled 0-Disabled ; Default = 0 means do not translate (don't make translation file(s)) and leave the original.
Enable=1
; Local translation file(s) auto-build. Based on Google Translate service.
; All original English-based (most often) submenus items will be tried to translate with dedicated async process tread.
; This means that translations will not appear immediately and are dependent on the operation and correct communication with Google Translate service (Min. timeout = 1 sec).
; Your local translation files will be created and systematically populated based on what you browse. After some time, YTuner will probably no longer need to query Google Translate.
; ! I don't take any responsibility for the correct operation of the Google Translate service !
; In case of an incorrect translation process, the values ​​in the dictionary may take values ​​such as !1!..!5! where the number between the "!" characters is the number of incorrect translation attempts (max = 5).
; 1-Enabled 0-Disabled ; Default: 0
Auto=1
; Translation target language code. Must be set if "Auto" is set to "1"!
; Valid values are here: https://cloud.google.com/translate/docs/languages
; Example : pl (Polish), de (German), es (Spanish), fr (French)
; Incorrect code will result in missing or incorrect translations
; Default: [blank] means : do not translate.
LanguageCode=pl
; Replace diacritic characters defined in "UTF-8 Latin1 Supplement" (https://www.w3schools.com/charsets/ref_utf_latin1_supplement.asp) with similar ANSI characters.
; Use this option if your AVR does not support your national character set included in the above "UTF-8 Latin1 Supplement" specification and displays strange characters like "_" instead.
; The character swap applies to submenu items and radio station names.
; This setting is independent of the above 'Enable' translator option!
; 1-Enabled 0-Disabled ; Default: 0
ReplaceUTF8Latin1Supplement=0
; Replace diacritic characters defined in "UTF-8 Latin Extended A" (https://www.w3schools.com/charsets/ref_utf_latin_extended_a.asp) with similar ANSI characters.
; Use this option if your AVR does not support your national character set included in the above "UTF-8 Latin Extended A" specification and displays strange characters like "_" instead.
; The character swap applies to submenu items and radio station names.
; This setting is independent of the above 'Enable' translator option!
; 1-Enabled 0-Disabled ; Default: 0
ReplaceUTF8Latin1ExtendedA=0
; Replace diacritic characters defined in "UTF-8 Latin Extended B" (https://www.w3schools.com/charsets/ref_utf_latin_extended_b.asp) with similar ANSI characters.
; Use this option if your AVR does not support your national character set included in the above "UTF-8 Latin Extended B" specification and displays strange characters like "_" instead.
; The character swap applies to submenu items and radio station names.
; This setting is independent of the above 'Enable' translator option!
; 1-Enabled 0-Disabled ; Default: 0
ReplaceUTF8Latin1ExtendedB=0
```

Below some most popular binaries of `YTuner v1.2.4 beta 1` to test:
[ytuner-1.2.4-beta1-x86_64-linux.zip](https://github.com/user-attachments/files/17792195/ytuner-1.2.4-beta1-x86_64-linux.zip)
[ytuner-1.2.4-beta1-armhf-linux.zip](https://github.com/user-attachments/files/17792197/ytuner-1.2.4-beta1-armhf-linux.zip)
[ytuner-1.2.4-beta1-aarch64-linux.zip](https://github.com/user-attachments/files/17792199/ytuner-1.2.4-beta1-aarch64-linux.zip)

Enjoy it :exclamation:  :speaker:  :notes:

**Reply by coffeegreg** [reply]
YTuner v1.2.4 has been released.

**Comment by coffeegreg** [comment]
Start of preparations for YTuner v1.2.5

Some minor fixes and improvements :
First try to fix : https://github.com/coffeegreg/YTuner/discussions/50
New option `IconEndPointExtension` in `ytuner.ini` file related to : https://github.com/coffeegreg/YTuner/discussions/43
>; This parameter is related to this discussion https://github.com/coffeegreg/YTuner/discussions/43
; This parameter can be used in very rare cases and in other cases it should be left empty.
; Default: blank.
; Example: jpg
IconEndPointExtension=

Fix : https://github.com/coffeegreg/YTuner/issues/44

@piet-t , @roland68 You should check this new version.

Below two most popular binaries of YTuner v1.2.5 alpha 1 to test:
[ytuner-1.2.5-alpha-1-x86_64-linux.zip](https://github.com/user-attachments/files/18548002/ytuner-1.2.5-alpha-1-x86_64-linux.zip)

**Reply by coffeegreg** [reply]
Second (final ?) try to fix : https://github.com/coffeegreg/YTuner/discussions/50

[ytuner-1.2.5-alpha-2-x86_64-linux.zip](https://github.com/user-attachments/files/18549684/ytuner-1.2.5-alpha-2-x86_64-linux.zip)

**Reply by roland68** [reply]
@coffeegreg
I have tested version 1.2.5, it works fine (#43 ).

**Reply by piet-t** [reply]
The final version works fine on my Harman Kardon including saving and recalling presets. Many thanks for the great and quick work.

**Reply by piet-t** [reply]
Sorry, I was a bit too quick. There's a minor quirk with recalling presets but it's also a strange traffic pattern that happens upon initial login.
From a UI point of view when switching to a stored preset most of the times the receiver switches channels but shows "empty" instead oof the station name on the display. Pressing "back" on the remote returns to the correct display. No idea what coulld be causing this or how to correct it, tuner's answer looks OK.
Sometimes it won't switch to the new channel at all but just show "empty" and won't react to any vtuner-specific buttons, only switching out of vtuner and back will get to a working state again.

From a network perspective, I see the following: when switching to a preset station the receiver seems to do a new login, then issue a search-query to harmankardon.vtuner.com - which never gets answered, instead the host closes the connection. Then the tuner issues the same request to harmankardon2.vtuner.com which somtimes gets an answer and sometimes doesn't.  Only successfully answered requests get a log-entry in ytuner's output, so the others seem to fail fairly early.

Regardless of the outcome of the first login/search-combo thee receiver always seems to issue another set of identical request right afterwards. In cases of the frst one failing and the second one succeeding the result is a successful switch to the new stream. If the first one succeeds and the second one fails I get to the situation where the receiver hangs,

I'll attach a zip with pcaps of both situations, perhaps you can find something in there.
--deleted wrong pdf file--

**Reply by coffeegreg** [reply]
Did you check this pdf attachment before attaching it? :smile: Sweet little baby ... :baby: 

First, try composing a console debug log from YTuner with a specific menu item pressed. This will help more in analyzing potential problems.
Schema:
Action1 : Pressing (some key)
Result1 : YTuner log lines
Action2 : Pressing (some key)
Result2 : YTuner log lines
... and so on ... marking where the problem occurred and what kind of problem. :+1: 
Let's also systematize the names... instead of "channel" let's use "station" or "radio station".

**Reply by piet-t** [reply]
Ooops, sorry about the attachment. This is what I wanted...
[recall_preset.zip](https://github.com/user-attachments/files/18551506/recall_preset.zip)

**Reply by coffeegreg** [reply]
Please make a schema like I wrote above and then we will be able to analyze potential problems. If such problems occur, they are likely not related to the changes introduced in version 1.2.5, but maybe their slightly different way of handling by HK AVR151S... so let's get back to the previous discussion thread -> https://github.com/coffeegreg/YTuner/discussions/50

**Comment by coffeegreg** [comment]
YTuner v1.2.5 has been released.

**Comment by cristiionita1989-cmd** [comment]
Does Ytuner suport harman hardon avr365? I'm instaling in Qnap TS433 and its runnibg with a dns error, loading the stations, but in the receiver i see ---empty--- in radio menu.

**Reply by TheBossME** [reply]
@cristiionita1989-cmd yes
